rust/src/main.rs

```diff
@@ -76,37 +109,147 @@ fn setup_logging() -> Result<(), fern::InitError> {
+    };
+    // Use the directory name set in logfile_config or fallback to the default
+    let logfile_directory_name = match logfile_config.directory {
+        Some(dir) => dir,
+        None => PathBuf::from(XI_LOG_DIR),
+    };
+
+    let mut logging_directory_path = get_logging_directory_path(logfile_directory_name)?;
+
+    // Add the file name & return the full path
+    logging_directory_path.push(logfile_file_name);
+    Ok(logging_directory_path)
+}
+
+fn get_flags() -> HashMap<String, Option<String>> {
+    let mut flags: HashMap<String, Option<String>> = HashMap::new();
+
+    let flag_prefix = "-";
+    let mut args_iterator = std::env::args().peekable();
+    while let Some(arg) = args_iterator.next() {
+        if arg.starts_with(flag_prefix) {
+            let key = arg.trim_left_matches(flag_prefix).to_string();
+
+            // Check the next argument doesn't start with the flag prefix
+            // map_or accounts for peek returning an Option
+            let next_arg_not_a_flag: bool = args_iterator
+                .peek()
+                .map_or(false, |val| !val.starts_with(flag_prefix));
+            if next_arg_not_a_flag {
+                flags.insert(key, args_iterator.next());
+            }
+        }
+    }
+    flags
+}
+
+struct EnvFlagConfig {
+    env_name: &'static str,
+    flag_name: &'static str,
+}
+
+/// Extracts a value from the flags and the env.
+///
+/// In this order: `String` from the flags, then `String` from the env, then `None`
+fn extract_env_or_flag(
+    flags: &HashMap<String, Option<String>>,
+    conf: EnvFlagConfig,
+) -> Option<String> {
+    flags
+        .get(conf.flag_name)
+        .cloned()
+        .unwrap_or(std::env::var(conf.env_name).ok())
+}
+
+struct LogfileConfig {
+    directory: Option<PathBuf>,
+    file: Option<PathBuf>,
+}
+
+fn generate_logfile_config(flags: &HashMap<String, Option<String>>) -> LogfileConfig {
+    // If the key is set, get the Option within
+    let log_dir_env_flag = EnvFlagConfig {
+        env_name: "XI_LOG_DIR",
+        flag_name: "log-dir",
+    };
+    let log_file_env_flag = EnvFlagConfig {
+        env_name: "XI_LOG_FILE",
+        flag_name: "log-file",
+    };
+    let log_dir_flag_option = extract_env_or_flag(&flags, log_dir_env_flag).map(PathBuf::from);
+
+    let log_file_flag_option = extract_env_or_flag(&flags, log_file_env_flag).map(PathBuf::from);
+
+    LogfileConfig {
+        directory: log_dir_flag_option,
+        file: log_file_flag_option,
+    }
+}
+
 fn main() {
     let mut state = XiCore::new();
     let stdin = io::stdin();
     let stdout = io::stdout();
     let mut rpc_looper = RpcLoop::new(stdout);
 
-    if let Err(e) = setup_logging() {
+    let flags: HashMap<String, Option<String>> = get_flags();
+
+    let logfile_config = generate_logfile_config(&flags);
+
+    let logging_path_result: Result<PathBuf, io::Error> = generate_logging_path(logfile_config);
+
+    let mut logging_path: Option<PathBuf> = None;
```

Comment:
this is confusing to me; it seems like either only one of these can be `None`, so why can't we keep this a `Result<PathBuf, io::Error>`? I suspect there's a reason but it sure looks odd.

If we _do_ need these to be a pair of optionals, it might be cleaner to match directly on `generate_logging_path(..)`

*some time has passed*

okay _still_ trying to figure out what's going on here. It sure looks like,

```rust
let log_path = generate_logging_path(logfile_config);
if let Err(err) = log_path.as_ref() {
    warn!("Failed to get log file path: '{}'", err);
}
if let Err(err) = setup_logging(log_path.ok()) {
    eprintln!("[Error] logging setup failed: {:?}, err);
}
```

Will do what you want, and is quite a bit (to me) clearer?

Reply:
It used to be that the `Result<PathBuf, io::Error>` was passed into `setup_logging` but it was probably cleaner to just give it an `Option<PathBuf>`.
The reason for the separate values is so the `warn!` could go after the `setup_logging` part.

Reply:
I just tried this code & the `warn!` bit didn't show up 😢 
```
let logging_path_result = Err(io::Error::new(
    io::ErrorKind::InvalidInput,
    "A blank file name was supplied",
));

if let Err(e) = logging_path_result.as_ref() {
    warn!(
        "Unable to generate the logging path to pass to set up: {}",
        e
    )
};
if let Err(e) = setup_logging(logging_path_result.ok()) {
    eprintln!(
        "[ERROR] setup_logging returned error, logging not enabled: {:?}",
        e
    );
};
```

Reply:
Ah right I understand. We could do the reverse just as easily, though,

```rust
let log_path = generate_logging_path(logfile_config);
if let Err(err) = setup_logging(log_path.as_ref().map(|p| p.as_path()).ok()) {
    eprintln!("[Error] logging setup failed: {:?}, err);
}
if let Err(err) = log_path {
    warn!("Failed to get log file path: '{}'", err);
}
```

in this version `setup_logging` takes an `Option<&Path>`.

Reply:
Works well.

```
let logging_path_result = Err(io::Error::new(io::ErrorKind::InvalidInput, "A blank file name was supplied"));

let logging_path = logging_path_result
    .as_ref()
    .map(|p: &PathBuf| -> &Path { p.as_path() })
    .ok();

if let Err(e) = setup_logging(logging_path) {
    eprintln!(
        "[ERROR] setup_logging returned error, logging not enabled: {:?}",
        e
    );
}
if let Err(e) = logging_path_result.as_ref() {
    warn!(
        "Unable to generate the logging path to pass to set up: {}",
        e
    )
};
```

I also fixed some stuff to work with the new `Option<&Path>`.

Reply:
I also like that `create_log_directory` is just using a `&Path` now.